### DISCUSSION TITLE
Internal error occurred during message handling | errno: -32

### DISCUSSION THREAD

**Discussion by mondial7** [enisdenjo/graphql-ws#437] [Q&A] [answered]
**What happens**

Few times a day we get this error, where some connections close unexpectedly. We are not really able to replicate it locally, and apparently the ws server does not crash/close.

**Open questions**

When in the process is such error triggered?
What does the `-32` error mean specifically?
Is there any way to get a more in depth stack trace / catch this failure?

**Full log**

```
Jan 17 02:28:04 PM  Internal error occurred during message handling. Please check your implementation. Error: write EPIPE
Jan 17 02:28:04 PM      at afterWriteDispatched (node:internal/stream_base_commons:160:15)
Jan 17 02:28:04 PM      at writevGeneric (node:internal/stream_base_commons:143:3)
Jan 17 02:28:04 PM      at Socket._writeGeneric (node:net:903:11)
Jan 17 02:28:04 PM      at Socket._writev (node:net:912:8)
Jan 17 02:28:04 PM      at doWrite (node:internal/streams/writable:408:12)
Jan 17 02:28:04 PM      at clearBuffer (node:internal/streams/writable:563:5)
Jan 17 02:28:04 PM      at Socket.Writable.uncork (node:internal/streams/writable:350:7)
Jan 17 02:28:04 PM      at Dh.sendFrame (/opt/[our-project-path]/build/index.js:1148:27327)
Jan 17 02:28:04 PM      at Dh.send (/opt/[our-project-path]/build/index.js:1148:26337)
Jan 17 02:28:04 PM      at fr.send (/opt/[our-project-path]/build/index.js:1148:36336) {
Jan 17 02:28:04 PM    errno: -32,
Jan 17 02:28:04 PM    code: 'EPIPE',
Jan 17 02:28:04 PM    syscall: 'write'
Jan 17 02:28:04 PM  }
```

**Packages**

`"graphql-ws": "~5.8.2"`
`"ws": "~8.6.0"`
Node: `16.14.0`

**Comment by enisdenjo** [comment, marked as answer]
Hey there, the error `write EPIPE` means that the program tried to write to a closed socket. We already prevent writing to a non-ready socket:

https://github.com/enisdenjo/graphql-ws/blob/eea18ce208abed12d3a9e61adc2c28d618cb3803/src/use/ws.ts#L135-L139

The error probably stems from the [`ws` library](https://github.com/websockets/ws) itself or some internal Node issue.

If you manage to reproduce this locally, I'd be more than happy to dig in and try to figure out the root cause! Sorry for the wait and if I was not very helpful.
